src/Parser.cpp

```diff
@@ -80,19 +81,57 @@ auto Parser::ParseDefinition() -> p<DefinitionAST> {
 }
 
 auto Parser::ParseRecordDef() -> p<DefinitionAST> {
-  auto record_tok = expect(TokRecord);
+  auto record_tok = expect(TokenType::TokRecord);
   if (!record_tok)
     return {nullptr, NONCOMMITTED};
-  auto id = expect(TokID);
 
+  auto id = expect(TokenType::TokID);
   if (!id) {
-
     this->add_error(record_tok->location,
                     "Failed to parse an identifier after token Record");
     return {nullptr, COMMITTED_NO_MORE_ERROR};
   }
-  // Julius continue from here :)
-  sammine_util::abort();
+
+  auto left_curly = expect(TokenType::TokLeftCurly);
+  if (!left_curly) {
+    this->add_error(record_tok->location,
+                    "Failed to parse an Left parenthesis after token Id");
+    return {nullptr, COMMITTED_NO_MORE_ERROR};
+  }
+
+  std::vector<std::unique_ptr<TypedVarAST>> record_members;
+  while (!tokStream->isEnd() &&
+         tokStream->peek()->tok_type != TokenType::TokRightCurly) {
+    auto [member, report] = ParseTypedVar();
+    if (report == SUCCESS) {
+      record_members.push_back(std::move(member));
+      if (!expect(TokenType::TokSemiColon)) {
+        this->add_error(record_tok->location,
+                        "Failed to parse an Semi colon after the Variable");
+        return {nullptr, COMMITTED_NO_MORE_ERROR};
+      }
+    }
+
+    else if (report == NONCOMMITTED) {
+      // probably end of the body
+      break;
+    } else {
+      this->add_error(record_tok->location, "Failed to parse record");
+      return {nullptr, COMMITTED_NO_MORE_ERROR};
```

Comment:
sorry, i know i leave a lot of reviews, but this is good to know i swear.

The sammine parser (and other production parser), despite the parser error, will always try to salvage even a partial AST. We'll do the same thing here. If you encounter an error here, you shouldn't return a nullptr because you'll make the AST printing stage lose valuable information on which sets of typed vars have we parsed successfully

To alleviate this, you should still construct the RecordDefAST with the partial vector of TypedVar that you got.

You'll notice that in the ParseProgramAST, the definition vector will push back whatever you return from ParseRecordDef

Reply:
so i should do something like?
```suggestion
      break;
    } else {
      this->add_error(record_tok->location, "Failed to parse record");
      auto rec = std::make_unique<RecordDefAST>(id, std::move(record_members));
      return {std::move(rec), COMMITTED_NO_MORE_ERROR};
```

Reply:
yeah looks good, you can even combine line 120 and 121 into the same line